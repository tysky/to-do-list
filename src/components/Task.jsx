import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import cn from 'classnames';
import * as actionCreators from '../actions';

const mapStateToProps = ({ tasks }) => {
  const props = {
    tasksList: Object.values(tasks),
  };
  return props;
};

class Task extends React.Component {
  state = { editing: false, editingText: '' }

  handleTaskCheckbox = () => {
    const { editTask, task: { id, status } } = this.props;
    const newStatus = status === 'todo' ? 'completed' : 'todo';
    editTask({ id, status: newStatus });
  }

  handleDeleteButton = () => {
    const {
      deleteTask, task: { id, taskIndex }, tasksList, editTask,
    } = this.props;
    deleteTask(id);
    tasksList.forEach((task) => {
      if (task.taskIndex > taskIndex) {
        editTask({ id: task.id, taskIndex: task.taskIndex - 1 });
      }
    });
  }

  handleEditButton = () => {
    const { task } = this.props;
    this.setState({ editing: true, editingText: task.text });
  }

  handleEditTaskText = (e) => {
    this.setState({ editingText: e.target.value });
  }

  editTask = (e) => {
    e.preventDefault();
    const { editTask, task } = this.props;
    const { editingText } = this.state;
    editTask({ id: task.id, text: editingText });
    this.setState({ editing: false });
  }

  renderEditTaskForm() {
    const { editingText } = this.state;
    return (
      <form onSubmit={this.editTask}>
        <input
          className="editTaskInput"
          type="text"
          required
          value={editingText}
          onChange={this.handleEditTaskText}
          autoComplete="off"
        />
        <button type="submit" className="edit">Edit</button>
      </form>
    );
  }

  render() {
    const { task } = this.props;
    const { editing } = this.state;
    const spanClass = cn({
      completed: task.status === 'completed',
    });
    return (
      <>
        <input
          type="checkbox"
          checked={task.status === 'completed'}
          onChange={this.handleTaskCheckbox}
        />
        {editing ? this.renderEditTaskForm()
          : <>
            <span className={spanClass}>{task.text}</span>
            <button type="button" className="edit" onClick={this.handleEditButton}>Edit</button>
           </>
        }
        <button type="button" className="delete" onClick={this.handleDeleteButton}>Delete</button>
      </>
    );
  }
}
Task.propTypes = {
  task: PropTypes.shape({
    id: PropTypes.number,
    text: PropTypes.string,
    status: PropTypes.string,
  }).isRequired,
  editTask: PropTypes.func.isRequired,
  deleteTask: PropTypes.func.isRequired,
  tasksList: PropTypes.arrayOf(PropTypes.object).isRequired,
};
export default connect(mapStateToProps, actionCreators)(Task);
