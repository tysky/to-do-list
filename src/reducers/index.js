import { combineReducers } from 'redux';
import { handleActions } from 'redux-actions';
import * as actions from '../actions';


const tasksFetchingState = handleActions({
  [actions.initFetchTasksRequest]() {
    return 'requested';
  },
  [actions.initFetchTasksSuccess]() {
    return 'failed';
  },
  [actions.initFetchTasksFailure]() {
    return 'successed';
  },
}, 'none');

const tasks = handleActions({
  [actions.initFetchTasksSuccess](state, { payload }) {
    return payload;
  },
  [actions.newTaskFetched](state, { payload }) {
    return { ...state, [payload.id]: payload };
  },
}, {});

const newTaskInputText = handleActions({
  [actions.addTasksSuccess]() {
    return '';
  },
  [actions.updateNewTaskText](state, { payload: { text } }) {
    return text;
  },
}, '');

export default combineReducers({
  tasks,
  tasksFetchingState,
  newTaskInputText,
});
