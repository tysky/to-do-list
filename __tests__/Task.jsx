import React from 'react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import Task from '../src/components/Task';
import reducers from '../src/reducers';

const task = {
  id: 1,
  status: 'todo',
  taskIndex: 1,
  text: 'task text',
};
const initialState = {
  tasks: {},
  tasksFetchingState: 'none',
  taskSendingState: 'none',
  newTaskInputText: '',
};

const store = createStore(reducers, initialState);
test('NewTaskInput', () => {
  const component = renderer.create(
    <Provider store={store}>
      <Task task={task} />
    </Provider>,
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
