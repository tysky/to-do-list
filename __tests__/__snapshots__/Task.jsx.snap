// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewTaskInput 1`] = `
Array [
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />,
  <span
    className=""
  >
    task text
  </span>,
  <button
    className="edit"
    onClick={[Function]}
    type="button"
  >
    Edit
  </button>,
  <button
    className="delete"
    onClick={[Function]}
    type="button"
  >
    Delete
  </button>,
]
`;
